backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/ReflectionSupport.kt

```diff
@@ -10,95 +10,165 @@ import org.jetbrains.kotlin.backend.konan.KonanBackendContext
 
-    private fun IrBuilderWithScope.irKTypeProjectionsList(
-            irTypeArguments: List<IrTypeArgument>
+    private fun IrBuilderWithScope.irKTypeParameter(
+            typeParameter: IrTypeParameter,
+            needExactTypeParameters: Boolean,
+            seenTypeParameters: MutableSet<IrTypeParameter>
     ): IrMemberAccessExpression<*> {
-        val kTypeProjectionType = symbols.kTypeProjection.typeWithoutArguments
-
-        return if (irTypeArguments.isEmpty()) {
-            irCall(symbols.emptyList, listOf(kTypeProjectionType))
-        } else {
-            irCall(symbols.listOf, listOf(kTypeProjectionType)).apply {
-                putValueArgument(0, IrVarargImpl(
-                        startOffset,
-                        endOffset,
-                        type = symbols.array.typeWith(kTypeProjectionType),
-                        varargElementType = kTypeProjectionType,
-                        elements = irTypeArguments.map { irKTypeProjection(it) }
-                ))
-            }
+        if (!seenTypeParameters.add(typeParameter))
+            throw RecursiveBoundsException("Non-reified type parameters with recursive bounds are not supported yet: ${typeParameter.render()}")
+        val result = irCall(symbols.kTypeParameterImpl.constructors.single()).apply {
+            putValueArgument(0, irString(typeParameter.name.asString()))
+            putValueArgument(1, irString(typeParameter.parentUniqueName))
+            putValueArgument(2, irKTypeList(typeParameter.superTypes, needExactTypeParameters, seenTypeParameters))
+            putValueArgument(3, irKVariance(typeParameter.variance))
+            putValueArgument(4, irBoolean(typeParameter.isReified))
         }
+        seenTypeParameters.remove(typeParameter)
+        return result
     }
 
-    private fun IrBuilderWithScope.irKTypeProjection(argument: IrTypeArgument): IrExpression {
-        return when (argument) {
-            is IrTypeProjection -> irCall(symbols.kTypeProjectionFactories.getValue(argument.variance)).apply {
-                dispatchReceiver = irGetObject(symbols.kTypeProjectionCompanion)
-                putValueArgument(0, irKType(argument.type))
-            }
+    private val IrTypeParameter.parentUniqueName get() = when (val parent = parent) {
+        is IrFunction -> parent.fullName
+        else -> parent.fqNameForIrSerialization.asString()
+    }
 
-            is IrStarProjection -> irCall(symbols.kTypeProjectionStar.owner.getter!!).apply {
-                dispatchReceiver = irGetObject(symbols.kTypeProjectionCompanion)
-            }
+    private fun IrBuilderWithScope.irKVariance(variance: Variance) =
+            IrGetEnumValueImpl(
+                    startOffset, endOffset,
+                    symbols.kVariance.defaultType,
+                    symbols.kVariance.owner.declarations.filterIsInstance<IrEnumEntry>()[variance.ordinal].symbol
```

Comment:
This approach may break for a number of reasons. Finding by name would be more reliable.